File: 100dynamic/2606.py
class Solution:
    def maximumCostSubstring(self, s: str, chars: str, vals: List[int]) -> int:
        va = [i for i in range(1,27)]

        for i in range(len(chars)):
            ch = chars[i]
            va[ord(ch) - ord('a')] = vals[i]
        arr = []
        for i in s:
            arr.append(va[ord(i) - ord('a')])
        
        def maxSubArray(nums: List[int]) -> int:
            n = len(nums)
            if (n == 1):
                return nums[0]
            tmp = nums[0]
            res = tmp
            for i in range(1,n):
                tmp = max(tmp + nums[i], nums[i])
                if tmp > res:
                    res = tmp
            return res
        res = maxSubArray(arr)
        return max(0,res)

File: 100dynamic/2606.rs
impl Solution {
    pub fn maximum_cost_substring(s: String, chars: String, vals: Vec<i32>) -> i32 {
         let mut va = [0;26];
         for i in 0..26 {
            va[i as usize] = i+1;
         }

        for (i,v) in chars.chars().enumerate() {
            va[v as usize - 'a' as usize] = vals[i]
        }

        let arr = s.chars().map(|c| va[c as usize - 'a' as usize]).collect();
        fn maxfunc(num: &Vec<i32>) -> i32 {
            let mut tmp = num[0];
            let mut res = num[0];
            for i in 1..num.len() {
                tmp = std::cmp::max(tmp + num[i], num[i]);
                if res < tmp {
                    res = tmp;
                }
            }
            res.max(0)
        }
        maxfunc(&arr)
    }
}